### PR TITLE
Import-VM-BlogPost

### PR DESCRIPTION
**Pull request by DirectedSoul1**
Signed-off-by: shegde <shegde@redhat.com>

### CODE REVIEW

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        lun: {}
+  volumes:
+    - name: mypvcdisk
+      persistentVolumeClaim:
+        claimName: mypvc
+```
+
+**Connect to VirtualMachineInstance console**
+
+Use virtctl to connect to the newly create VirtualMachinInstance.
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: 'Machine'

**Comment by iranzo** [comment]
** is used instead of ropper headings "##" please, replace what is a header with  proper heading level, and leave ** for things like notes, etc

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+      volumes:
+      - name: containerdisk
+        containerDisk:
+          image: kubevirt/cirros-container-disk-demo:latest
+      - cloudInitNoCloud:
+          userDataBase64: IyEvYmluL3NoCgplY2hvICdwcmludGVkIGZyb20gY2xvdWQtaW5pdCB1c2VyZGF0YScK
+        name: cloudinitdisk
+
+```
+**Note**:
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: There's a new note very close, should those be merged together?

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        name: cloudinitdisk
+
+```
+**Note**: 
+From the above manifest, `kind: VirtualMachine` states that its a VM object, `spec.domain.device.name` section and `spec.volumes.name` should match. In the later section of this Blog you will see how this is all gets connected in the context of CDI.
+
+**Note**:
+
+- More examples of a VM declared as a `yaml` manifest can be seen [here](https://github.com/kubevirt/kubevirt/tree/master/cluster/examples)
+
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Remove spaces between "-" items

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+A VirtualMachine will make sure that a VirtualMachineInstance object with an identical name will be present in the cluster, if `spec.running` is set to `true`. Further it will make sure that a VirtualMachineInstance will be removed from the cluster if `spec.running` is set to `false`.
+
+There exists a field `spec.runStrategy` which can also be used to control the state of the associated VirtualMachineInstance object. To avoid confusing and contradictory states, these fields are mutually exclusive. An extended explanation of `spec.runStrategy` vs `spec.running` can be found in [Run Strategies](https://kubernetes.io/docs/tasks/run-application/run-single-instance-stateful-application/)
+
+Saving this manifest into vm.yaml and submitting it to Kubernetes will create the controller instance:
+
+```shell
+$ kubectl create -f vm.yaml
+virtualmachine "vm-cirros" created
+```
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Leave extra new line between code blocks and next paragraph and same for others before it

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+Since spec.running is set to false, no vmi will be created:
+
+```shell
+$ kubectl get vmis
+No resources found.
+```
+Let’s start the VirtualMachine:
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: newline before

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+kubectl create -f https://github.com/kubevirt/containerized-data-importer/releases/download/$VERSION/cdi-operator.yaml
+kubectl create -f https://github.com/kubevirt/containerized-data-importer/releases/download/$VERSION/cdi-operator-cr.yaml
+```
+
+**Expose cdi-uploadproxy service:**
+
+The cdi-uploadproxy service must be accessible from outside the cluster. Here are some ways to do that:
+
+- [NodePort Service](https://kubernetes.io/docs/concepts/services-networking/service/#nodeport)
+
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Remove extra newline

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+- [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/)
+
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Same here

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+- [Route](https://docs.openshift.com/container-platform/3.9/architecture/networking/routes.html)
+
+We can take a look at example manifests [here](https://github.com/kubevirt/containerized-data-importer/blob/master/doc/upload.md)
+
+The supported image formats are:
+
+- .img
```
Comment (iranzo) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Use codeblocks for extension and add in 238 the comment about compressed

**Comment by iranzo** [comment]
In general, I see that this PR is not based on latest suggestions/changes performed on the previous PR, please, do check them as starting point.

Thanks!

**Comment by karmab** [comment]
can you address the issues @DirectedSoul1  ?

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 1 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Added the chnges :
```
Use virtctl to connect to the newly create VirtualMachineInstance.
```

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 2 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Yes, very good suggestion, I have merged both the notes into one.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 4 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Added the new lines.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 6 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Removed the new-lines.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 7 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Fixed the new-line spaces.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 8 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Added code Blocks.
```
- .img

- .iso

- .qcow2

- Compressed .tar, .gz and .xz of the above formats are supported.
```

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 3 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Removed the new Lines.
```
- More examples of a VM declared as a `yaml` manifest can be seen [here](https://github.com/kubevirt/kubevirt/tree/master/cluster/examples)
- Please feel free to take a look at how to deploy VM as a K8s object by using kubevirt add-on using minikube [here](https://kubevirt.io//quickstart_minikube/)
```

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 5 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Added the new line.
```
$ kubectl get vmis
No resources found.
```

Let’s start the VirtualMachine:

**Comment by DirectedSoul1** [comment]
> ** is used instead of ropper headings "##" please, replace what is a header with proper heading level, and leave ** for things like notes, etc

Added the ## and removed the "**" for the headings.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+description: Import a VM into the Kubernetes Platform using CDI
+navbar_active: Blogs
+pub-date: May 20
+pub-year: 2019
+category: news
+---
+
+# Importing a VM into the Kubernetes Native Environment
+
+Motivation: Kubernetes have become the new way to orchestrate the containers and to handle the microservice architecture, but what if I already have legacy apps running on my old VM's in my datacenter ? Can those apps ever be made k8s friendly ? Well if that is the usecase for you then we have a solution now with kubevirt.
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: - kubernetes has become
- and handle the microservices
i wouldnt use the term legacy, just apps (or older apps)

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+
+In this blog post we will discuss about how to deploy VM as a yaml template and steps on how to import it as a [PVC](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) onto your kubernetes environment using the CDI and kubevirt add-ons.
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: we will show how to deploy a vm

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+
+**Assumptions:**
+
+- A basic understanding of the k8s architecture: In its simplest terms Kubernetes is a portable, extensible open-source platform for managing containerized workloads and services, that facilitates both declarative configuration and automation. It has a large, rapidly growing ecosystem. Kubernetes services, support, and tools are widely available. For complete details check [Kubernetes-architecture](https://www.aquasec.com/wiki/display/containers/Kubernetes+Architecture+101)
+
+- User is familiar with the concept of a [virsh based VM](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_deployment_and_administration_guide/sect-guest_virtual_machine_installation_overview-creating_guests_with_virt_install)
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: of a libvirt based vm, rather ?

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+
+- PersistentVolume (PV) is a piece of storage in the cluster that has been provisioned by an administrator. It is a resource in the cluster just like a node is a cluster resource. PVs are volume plugins like Volumes, but have a lifecycle independent of any individual pod that uses the PV. This API object captures the details of the implementation of the storage, be that NFS, iSCSI, or a cloud-provider-specific storage system. [Persistent Volume(PV)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and Persistent Volume Claim PersistentVolumeClaim (PVC) is a request for storage by a user. It is similar to a pod. Pods consume node resources and PVCs consume PV resources. Pods can request specific levels of resources (CPU and Memory). Claims can request specific size and access modes (e.g., can be mounted once read/write or many times read-only). [Persistent Volume Claim(PVC)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims).
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: i dont think we need to explain what a pv/pvc is, worst case a single sentence and a link to the documentation

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+
+- User is familiar with the concept of [kubevirt-architecture](https://github.com/kubevirt/kubevirt/blob/master/docs/architecture.md) and [CDI-architecture](https://github.com/kubevirt/containerized-data-importer/blob/master/doc/design.md#design)
+
+- User has already installed kubevirt in his K8s environment, if not please follow the link [Installing Kubevirt](https://kubevirt.io/user-guide/docs/latest/administration/intro.html#installation)
+
+# Virtual machines:
+
+**Virtual machines:**
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: i believe we dont need to explain (again) what a vm does and how to handle it, user guide contains enough information

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+
+For detailed usage of Volumes, you can take a look at [here](https://kubevirt.io/user-guide/docs/latest/creating-virtual-machines/disks-and-volumes.html)
+
+Now, we have seen how a VM can be declared as a yaml object inside kubernetes, we will now see how CDI and kubevirt help in getting our VM workload into the cluster.Next immediate question that comes to our mind is "How to use a Virtual Machine in my k8s cluster?", ok, lets see how all of this is made possible.
+
+A VirtualMachine will make sure that a VirtualMachineInstance object with an identical name will be present in the cluster, if `spec.running` is set to `true`. Further it will make sure that a VirtualMachineInstance will be removed from the cluster if `spec.running` is set to `false`.
+
+There exists a field `spec.runStrategy` which can also be used to control the state of the associated VirtualMachineInstance object. To avoid confusing and contradictory states, these fields are mutually exclusive. An extended explanation of `spec.runStrategy` vs `spec.running` can be found in [Run Strategies](https://kubernetes.io/docs/tasks/run-application/run-single-instance-stateful-application/)
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: is this field especially relevant in the context of importing a vm. if yes, let s indicate why

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+
+- Import a disk image from a URL to a PVC (HTTP/S3)
+
+- Clone an an existing PVC
+
+- Upload a local disk image to a PVC
+
+This document for now deals with the third use case and covers the HTTP based import use case at the end of the BlogPost. So you should have CDI installed in your cluster, a VM disk that you’d like to upload, and virtctl in your path.
+
+Lets begin by installing the latest CDI release [here](https://github.com/kubevirt/containerized-data-importer/releases) (currently v1.9.0)
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: i think we dont need to indicate how to install cdi, it s documented on cdi web page

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+
+Use virtctl to connect to the newly created VirtualMachineInstance.
+
+```shell
+virtctl console cirros-vm
+```
+
+The above method explains the import of the VM by importing the locally available image, now lets see the concept of DataVolumes:
+
+# DataVolume:
```
Comment (karmab) [inline _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: does using datavolumes improve the importing process. if yes, let s outline why. If not, i dont know if we want to talk about data volumes

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 12 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Taken care of( by putting a single sentence and an extended link.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 14 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: It's not really important in the context of VM import, but it does explain as to how the VMI itself gets started in the kubevirt environment. Should this be removed?

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 15 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Makes sense, removed the paragraph about CDI explanation. But I have pointed the reader to the GitHub link of CDI.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 9 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Made the changes as suggested. Thanks for pointing this out @karmab

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 10 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Edited as suggested.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 11 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Made the changes.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 13 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Are you suggesting to remove all of these :
```
# Virtual machines:

**Virtual machines:**
A VirtualMachine provides additional management capabilities to a VirtualMachineInstance inside the cluster. That includes:

- ABI stability

- Start/stop/restart capabilities on the controller level

- Offline configuration change with propagation on VirtualMachineInstance recreation

- Ensure that the VirtualMachineInstance is running if it should be running

It focuses on a 1:1 relationship between the controller instance and a virtual machine instance. In many ways it is very similar to a StatefulSet with spec.replica set to 1.
```

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 16 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: It does not improve the process, but it gives a good idea of the whole `yaml` structure, i do feel to get a complete picture this section is needed.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 14 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: removed the `spec.runStrategy` field, as it is not really needed in this context.

File: _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown (hunk 13 above)
Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Added a one-liner and pointed an external link for more.

Comment (DirectedSoul1) [reply _posts/2019-05-15-How-To-Guide-Import-VM-Kubevirt.markdown]: Deleted non-relevant concepts.

**Comment by iranzo** [comment]
Rebase this one too to fix the issue with the link that is fixed in master please

**Pull request merged**